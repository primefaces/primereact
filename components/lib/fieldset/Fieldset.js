import React, { forwardRef, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { Ripple } from '../ripple/Ripple';
import { CSSTransition } from '../csstransition/CSSTransition';
import { classNames, UniqueComponentId } from '../utils/Utils';
import { useMountEffect } from '../hooks/Hooks';

export const Fieldset = forwardRef((props, ref) => {
    const [idState, setIdState] = useState(props.id);
    const [collapsedState, setCollapsedState] = useState(props.collapsed);
    const collapsed = props.toggleable ? (props.onToggle ? props.collapsed : collapsedState) : false;
    const contentRef = useRef(null);
    const headerId = idState + '_header';
    const contentId = idState + '_content';

    const toggle = (event) => {
        if (props.toggleable) {
            collapsed ? expand(event) : collapse(event);

            if (props.onToggle) {
                props.onToggle({
                    originalEvent: event,
                    value: !collapsed
                });
            }
        }

        event.preventDefault();
    }

    const expand = (event) => {
        if (!props.onToggle) {
            setCollapsedState(false);
        }

        props.onExpand && props.onExpand(event);
    }

    const collapse = (event) => {
        if (!props.onToggle) {
            setCollapsedState(true);
        }

        props.onCollapse && props.onCollapse(event);
    }

    useMountEffect(() => {
        if (!props.id) {
            setIdState(UniqueComponentId());
        }
    })

    const createContent = () => {
        return (
            <CSSTransition nodeRef={contentRef} classNames="p-toggleable-content" timeout={{ enter: 1000, exit: 450 }} in={!collapsed} unmountOnExit options={props.transitionOptions}>
                <div ref={contentRef} id={contentId} className="p-toggleable-content" aria-hidden={collapsed} role="region" aria-labelledby={headerId}>
                    <div className="p-fieldset-content">
                        {props.children}
                    </div>
                </div>
            </CSSTransition>
        )
    }

    const createToggleIcon = () => {
        if (props.toggleable) {
            const className = classNames('p-fieldset-toggler pi', {
                'pi-plus': collapsed,
                'pi-minus': !collapsed
            });

            return <span className={className}></span>
        }

        return null;
    }

    const createLegendContent = () => {
        if (props.toggleable) {
            const toggleIcon = createToggleIcon();

            return (
                <a href={'#' + contentId} aria-controls={contentId} id={headerId} aria-expanded={!collapsed} tabIndex={props.toggleable ? null : -1}>
                    {toggleIcon}
                    <span className="p-fieldset-legend-text">{props.legend}</span>
                    <Ripple />
                </a>
            )
        }

        return <span className="p-fieldset-legend-text" id={headerId}>{props.legend}</span>
    }

    const createLegend = () => {
        if (props.legend != null || props.toggleable) {
            const legendContent = createLegendContent();

            return (
                <legend className="p-fieldset-legend p-unselectable-text" onClick={toggle}>
                    {legendContent}
                </legend>
            )
        }
    }

    const className = classNames('p-fieldset p-component', {
        'p-fieldset-toggleable': props.toggleable
    }, props.className);
    const legend = createLegend();
    const content = createContent();

    return (
        <fieldset id={idState} className={className} style={props.style} onClick={props.onClick}>
            {legend}
            {content}
        </fieldset>
    )
});

Fieldset.defaultProps = {
    __TYPE: 'Fieldset',
    id: null,
    legend: null,
    className: null,
    style: null,
    toggleable: null,
    collapsed: null,
    transitionOptions: null,
    onExpand: null,
    onCollapse: null,
    onToggle: null,
    onClick: null
};

Fieldset.propTypes /* remove-proptypes */ = {
    __TYPE: PropTypes.string,
    id: PropTypes.string,
    legend: PropTypes.any,
    className: PropTypes.string,
    style: PropTypes.object,
    toggleable: PropTypes.bool,
    collapsed: PropTypes.bool,
    transitionOptions: PropTypes.object,
    onExpand: PropTypes.func,
    onCollapse: PropTypes.func,
    onToggle: PropTypes.func,
    onClick: PropTypes.func
};
