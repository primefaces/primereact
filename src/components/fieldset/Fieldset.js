import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { classNames } from '../utils/ClassNames';
import UniqueComponentId from '../utils/UniqueComponentId';
import { CSSTransition } from '../transition/CSSTransition';
import { Ripple } from '../ripple/Ripple';

export class Fieldset extends Component {

    static defaultProps = {
        id: null,
        legend: null,
        className: null,
        style: null,
        toggleable: null,
        collapsed: null,
        transitionOptions: null,
        onExpand: null,
        onCollapse: null,
        onToggle: null,
        onClick: null
    };

    static propTypes = {
        id: PropTypes.string,
        legend: PropTypes.any,
        className: PropTypes.string,
        style: PropTypes.object,
        toggleable: PropTypes.bool,
        collapsed: PropTypes.bool,
        transitionOptions: PropTypes.object,
        onExpand: PropTypes.func,
        onCollapse: PropTypes.func,
        onToggle: PropTypes.func,
        onClick: PropTypes.func
    };

    constructor(props) {
        super(props);
        let state = {
            id: props.id
        };

        if (!this.props.onToggle) {
            state = {
                ...state,
                collapsed: props.collapsed
            };
        }

        this.state = state;

        this.toggle = this.toggle.bind(this);
        this.contentRef = React.createRef();
    }

    toggle(event) {
        if (this.props.toggleable) {
            const collapsed = this.props.onToggle ? this.props.collapsed : this.state.collapsed;

            if (collapsed)
                this.expand(event);
            else
                this.collapse(event);

            if (this.props.onToggle) {
                this.props.onToggle({
                    originalEvent: event,
                    value: !collapsed
                });
            }
        }

        event.preventDefault();
    }

    expand(event) {
        if (!this.props.onToggle) {
            this.setState({ collapsed: false });
        }

        if (this.props.onExpand) {
            this.props.onExpand(event);
        }
    }

    collapse(event) {
        if (!this.props.onToggle) {
            this.setState({ collapsed: true });
        }

        if (this.props.onCollapse) {
            this.props.onCollapse(event);
        }
    }

    isCollapsed() {
        return this.props.toggleable ? (this.props.onToggle ? this.props.collapsed : this.state.collapsed) : false;
    }

    componentDidMount() {
        if (!this.state.id) {
            this.setState({ id: UniqueComponentId() });
        }
    }

    renderContent(collapsed) {
        const id = this.state.id + '_content';

        return (
            <CSSTransition nodeRef={this.contentRef} classNames="p-toggleable-content" timeout={{ enter: 1000, exit: 450 }} in={!collapsed} unmountOnExit options={this.props.transitionOptions}>
                <div ref={this.contentRef} id={id} className="p-toggleable-content" aria-hidden={collapsed} role="region" aria-labelledby={this.state.id + '_header'}>
                    <div className="p-fieldset-content">
                        {this.props.children}
                    </div>
                </div>
            </CSSTransition>
        );
    }

    renderToggleIcon(collapsed) {
        if (this.props.toggleable) {
            const className = classNames('p-fieldset-toggler pi', { 'pi-plus': collapsed, 'pi-minus': !collapsed });

            return (
                <span className={className}></span>
            );
        }

        return null;
    }

    renderLegendContent(collapsed) {
        if (this.props.toggleable) {
            const toggleIcon = this.renderToggleIcon(collapsed);
            const ariaControls = this.state.id + '_content';

            return (
                <a href={'#' + ariaControls} aria-controls={ariaControls} id={this.state.id + '_header'} aria-expanded={!collapsed} tabIndex={this.props.toggleable ? null : -1}>
                    {toggleIcon}
                    <span className="p-fieldset-legend-text">{this.props.legend}</span>
                    <Ripple />
                </a>
            );
        }

        return (
            <span className="p-fieldset-legend-text" id={this.state.id + '_header'}>{this.props.legend}</span>
        );
    }

    renderLegend(collapsed) {
        const legendContent = this.renderLegendContent(collapsed);
        if (this.props.legend != null || this.props.toggleable) {
            return (
                <legend className="p-fieldset-legend p-unselectable-text" onClick={this.toggle}>
                    {legendContent}
                </legend>
            );
        }
    }

    render() {
        const className = classNames('p-fieldset p-component', this.props.className, { 'p-fieldset-toggleable': this.props.toggleable });
        const collapsed = this.isCollapsed();
        const legend = this.renderLegend(collapsed);
        const content = this.renderContent(collapsed);

        return (
            <fieldset id={this.props.id} className={className} style={this.props.style} onClick={this.props.onClick}>
                {legend}
                {content}
            </fieldset>
        );
    }
}
