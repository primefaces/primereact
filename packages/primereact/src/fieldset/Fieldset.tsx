'use client';
import { Component } from '@primereact/core/component';
import { useFieldset } from '@primereact/headless/fieldset';
import { styles } from '@primereact/styles/fieldset';
import { mergeProps } from '@primeuix/utils';
import { withComponent } from 'primereact/base';
import * as React from 'react';
import { FieldsetContent } from './content';
import { FieldsetProvider } from './Fieldset.context';
import { defaultProps } from './Fieldset.props';
import { FieldsetLegend } from './legend';

export const Fieldset = withComponent({
    name: 'Fieldset',
    defaultProps,
    styles,
    setup(instance) {
        const fieldset = useFieldset(instance.inProps);

        return fieldset;
    },
    render(instance) {
        const { id, props, ptmi, cx } = instance;

        const rootProps = mergeProps(
            {
                id,
                className: cx('root')
            },
            ptmi('root')
        );

        return (
            <FieldsetProvider value={instance}>
                <Component instance={instance} attrs={rootProps} children={props.children} />
            </FieldsetProvider>
        );
    },
    components: {
        Legend: FieldsetLegend,
        Content: FieldsetContent
    }
});
