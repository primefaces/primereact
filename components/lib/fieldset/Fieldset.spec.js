import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { snapshot } from '../../test';
import { Fieldset } from './Fieldset';

describe('Fieldset', () => {
    snapshot(<Fieldset />, 'default');
    snapshot(
        <Fieldset id="fieldset" legend="Simple Fieldset">
            Content
        </Fieldset>,
        'legend'
    );
    test('when Fieldset is toggleable it will toggle when clicked', async () => {
        // Arrange
        const toggleOn = jest.fn();
        const { container } = render(
            <Fieldset legend="Toggleable" toggleable onToggle={toggleOn}>
                <p>Lorem ipsum dolor sit amet</p>
            </Fieldset>
        );
        const legend = container.getElementsByClassName('p-fieldset-legend')[0];

        expect(container).toMatchSnapshot('toggleable-open');

        // Act
        await userEvent.click(legend);

        // Assert
        expect(toggleOn).toHaveBeenCalledTimes(1);
        expect(container).toMatchSnapshot('toggleable-closed');
    });
    test('when Fieldset is toggleable it must expand and collapse', async () => {
        // Arrange
        const expandOn = jest.fn();
        const collapseOn = jest.fn();
        const { container } = render(
            <Fieldset legend="Expand/Collapse" toggleable onExpand={expandOn} onCollapse={collapseOn}>
                <p>Lorem ipsum dolor sit amet</p>
            </Fieldset>
        );
        const legend = container.getElementsByClassName('p-fieldset-legend')[0];

        expect(container).toMatchSnapshot('expandable-open');

        // Act
        await userEvent.click(legend);

        // Assert
        expect(expandOn).toHaveBeenCalledTimes(0);
        expect(collapseOn).toHaveBeenCalledTimes(1);
        expect(container).toMatchSnapshot('expandable-closed');

        // Act
        await userEvent.click(legend);

        // Assert
        expect(expandOn).toHaveBeenCalledTimes(1);
        expect(collapseOn).toHaveBeenCalledTimes(1);
        expect(container).toMatchSnapshot('expandable-open');
    });
});
