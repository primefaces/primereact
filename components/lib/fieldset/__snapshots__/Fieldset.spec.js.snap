// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fieldset default 1`] = `
<div>
  <fieldset
    class="p-fieldset p-component"
    id="pr_id_1"
  >
    <div
      aria-hidden="false"
      aria-labelledby="pr_id_1_header"
      class="p-toggleable-content"
      id="pr_id_1_content"
      role="region"
    >
      <div
        class="p-fieldset-content"
      />
    </div>
  </fieldset>
</div>
`;

exports[`Fieldset legend 1`] = `
<div>
  <fieldset
    class="p-fieldset p-component"
    id="fieldset"
  >
    <legend
      class="p-fieldset-legend p-unselectable-text"
    >
      <span
        class="p-fieldset-legend-text"
        id="fieldset_header"
      >
        Simple Fieldset
      </span>
    </legend>
    <div
      aria-hidden="false"
      aria-labelledby="fieldset_header"
      class="p-toggleable-content"
      id="fieldset_content"
      role="region"
    >
      <div
        class="p-fieldset-content"
      >
        Content
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`Fieldset when Fieldset is toggleable it must expand and collapse: expandable-closed 1`] = `
<div>
  <fieldset
    class="p-fieldset p-component p-fieldset-toggleable"
    id="pr_id_3"
  >
    <legend
      class="p-fieldset-legend p-unselectable-text"
    >
      <a
        aria-controls="pr_id_3_content"
        aria-expanded="false"
        href="#pr_id_3_content"
        id="pr_id_3_header"
      >
        <span
          class="p-fieldset-toggler pi pi-plus"
        />
        <span
          class="p-fieldset-legend-text"
        >
          Expand/Collapse
        </span>
      </a>
    </legend>
    <div
      aria-hidden="true"
      aria-labelledby="pr_id_3_header"
      class="p-toggleable-content p-toggleable-content-exit p-toggleable-content-exit-active"
      id="pr_id_3_content"
      role="region"
    >
      <div
        class="p-fieldset-content"
      >
        <p>
          Lorem ipsum dolor sit amet
        </p>
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`Fieldset when Fieldset is toggleable it must expand and collapse: expandable-open 1`] = `
<div>
  <fieldset
    class="p-fieldset p-component p-fieldset-toggleable"
    id="pr_id_3"
  >
    <legend
      class="p-fieldset-legend p-unselectable-text"
    >
      <a
        aria-controls="pr_id_3_content"
        aria-expanded="true"
        href="#pr_id_3_content"
        id="pr_id_3_header"
      >
        <span
          class="p-fieldset-toggler pi pi-minus"
        />
        <span
          class="p-fieldset-legend-text"
        >
          Expand/Collapse
        </span>
      </a>
    </legend>
    <div
      aria-labelledby="pr_id_3_header"
      class="p-toggleable-content"
      id="pr_id_3_content"
      role="region"
    >
      <div
        class="p-fieldset-content"
      >
        <p>
          Lorem ipsum dolor sit amet
        </p>
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`Fieldset when Fieldset is toggleable it must expand and collapse: expandable-open 2`] = `
<div>
  <fieldset
    class="p-fieldset p-component p-fieldset-toggleable"
    id="pr_id_3"
  >
    <legend
      class="p-fieldset-legend p-unselectable-text"
    >
      <a
        aria-controls="pr_id_3_content"
        aria-expanded="true"
        href="#pr_id_3_content"
        id="pr_id_3_header"
      >
        <span
          class="p-fieldset-toggler pi pi-minus"
        />
        <span
          class="p-fieldset-legend-text"
        >
          Expand/Collapse
        </span>
      </a>
    </legend>
    <div
      aria-hidden="false"
      aria-labelledby="pr_id_3_header"
      class="p-toggleable-content p-toggleable-content-enter p-toggleable-content-enter-active"
      id="pr_id_3_content"
      role="region"
    >
      <div
        class="p-fieldset-content"
      >
        <p>
          Lorem ipsum dolor sit amet
        </p>
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`Fieldset when Fieldset is toggleable it will toggle when clicked: toggleable-closed 1`] = `
<div>
  <fieldset
    class="p-fieldset p-component p-fieldset-toggleable"
    id="pr_id_2"
  >
    <legend
      class="p-fieldset-legend p-unselectable-text"
    >
      <a
        aria-controls="pr_id_2_content"
        aria-expanded="true"
        href="#pr_id_2_content"
        id="pr_id_2_header"
      >
        <span
          class="p-fieldset-toggler pi pi-minus"
        />
        <span
          class="p-fieldset-legend-text"
        >
          Toggleable
        </span>
      </a>
    </legend>
    <div
      aria-labelledby="pr_id_2_header"
      class="p-toggleable-content"
      id="pr_id_2_content"
      role="region"
    >
      <div
        class="p-fieldset-content"
      >
        <p>
          Lorem ipsum dolor sit amet
        </p>
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`Fieldset when Fieldset is toggleable it will toggle when clicked: toggleable-open 1`] = `
<div>
  <fieldset
    class="p-fieldset p-component p-fieldset-toggleable"
    id="pr_id_2"
  >
    <legend
      class="p-fieldset-legend p-unselectable-text"
    >
      <a
        aria-controls="pr_id_2_content"
        aria-expanded="true"
        href="#pr_id_2_content"
        id="pr_id_2_header"
      >
        <span
          class="p-fieldset-toggler pi pi-minus"
        />
        <span
          class="p-fieldset-legend-text"
        >
          Toggleable
        </span>
      </a>
    </legend>
    <div
      aria-labelledby="pr_id_2_header"
      class="p-toggleable-content"
      id="pr_id_2_content"
      role="region"
    >
      <div
        class="p-fieldset-content"
      >
        <p>
          Lorem ipsum dolor sit amet
        </p>
      </div>
    </div>
  </fieldset>
</div>
`;
